cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/xcm_fee_estimation.rs

```diff
@@ -0,0 +1,373 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Tests to ensure correct XCM fee estimation between Asset Hub Westend and the Westend relay
+//! chain.
+
+use crate::imports::*;
+
+use sp_keyring::AccountKeyring::Alice;
+use sp_runtime::{generic, MultiSignature};
+use xcm_fee_payment_runtime_api::{
+	dry_run::runtime_decl_for_xcm_dry_run_api::XcmDryRunApiV1,
+	fees::runtime_decl_for_xcm_payment_api::XcmPaymentApiV1,
+};
+
+type RelayToAssetHubTest = Test<Westend, AssetHubWestend>;
+
+/// We are able to dry-run and estimate the fees for a teleport between relay and system para.
+/// Scenario: Alice on Westend relay chain wants to teleport WND to Asset Hub.
+/// We want to know the fees using the `XcmDryRunApi` and `XcmPaymentApi`.
+#[test]
+fn teleport_relay_system_para_works() {
+	let destination: Location = Parachain(1000).into(); // Asset Hub.
+	let beneficiary_id = AssetHubWestendReceiver::get();
+	let beneficiary: Location = AccountId32 { id: beneficiary_id.clone().into(), network: None } // Test doesn't allow specifying a network here.
+		.into(); // Beneficiary in Asset Hub.
+	let teleport_amount = 1_000_000_000_000; // One WND (12 decimals).
+	let assets: Assets = vec![(Here, teleport_amount).into()].into();
+
+	// We get them from the Westend closure.
+	let mut delivery_fees_amount = 0;
+	let mut remote_message = VersionedXcm::V4(Xcm(Vec::new()));
+	<Westend as TestExt>::new_ext().execute_with(|| {
+		type Runtime = <Westend as Chain>::Runtime;
+		type RuntimeCall = <Westend as Chain>::RuntimeCall;
+
+		let call = RuntimeCall::XcmPallet(pallet_xcm::Call::transfer_assets {
+			dest: Box::new(VersionedLocation::V4(destination.clone())),
+			beneficiary: Box::new(VersionedLocation::V4(beneficiary)),
+			assets: Box::new(VersionedAssets::V4(assets)),
+			fee_asset_item: 0,
+			weight_limit: Unlimited,
+		});
+		let sender = Alice; // Is the same as `WestendSender`.
+		let extrinsic = construct_extrinsic_westend(sender, call);
+		let result = Runtime::dry_run_extrinsic(extrinsic).unwrap();
+		let (destination_to_query, messages_to_query) = &result.forwarded_xcms[0];
+		remote_message = messages_to_query[0].clone();
+		let delivery_fees =
+			Runtime::query_delivery_fees(destination_to_query.clone(), remote_message.clone())
+				.unwrap();
+		delivery_fees_amount = get_amount_from_versioned_assets(delivery_fees);
+		assert_eq!(delivery_fees_amount, 39_700_000_000);
```

Comment:
Let's not check exact amounts as they will constantly change.
```suggestion
```

The final check "received amount = sent - fees" covers everything we want 👍